lib/vast/Tower/Link.cpp

```diff
@@ -0,0 +1,171 @@
+        op->walk(collect);
+        return out;
+    }
+
+    std::vector< op_mapping > mk_link_mappings(const auto &links) {
+        std::vector< op_mapping > out;
+        for (const auto &l : links)
+            out.emplace_back(l->parents_to_children());
+        return out;
+    }
+
+    // `transition` handles the lookup between levels - we need this as we want to generalise
+    // (`levels` can be arbitrary mapping, not just `op -> { op }`).
+    op_mapping build_map(op_mapping init, const auto &levels, auto transition) {
+        auto handle_level = [&](const auto &level) {
+            auto handle_element = [&](operation op) -> operations {
+                auto prev_ops = level.find(transition(op));
+                if (prev_ops == level.end())
+                    return {};
+                return prev_ops->second;
+            };
+
+            for (auto &[op, todo] : init) {
+                operations parents;
+                for (auto current : todo) {
+                    parents.append_range(handle_element(current));
+                }
+                todo = std::move(parents);
+            }
+        };
+
+        for (const auto &level : levels | std::views::drop(1))
+            handle_level(level);
+
+        return init;
+    }
+
+    op_mapping build_map(const link_vector &links) {
+        auto transition = [](operation op) { return op; };
+        auto init = links.front()->parents_to_children();
+        return build_map(init, mk_link_mappings(links), transition);
+    }
+
+    op_mapping build_map(std::vector< loc_to_op_t > links, location_info_t &li) {
+        auto transition = [&](operation op) { return li.prev(op); };
+
+        std::reverse(links.begin(), links.end());
+
+        // initialize
+        op_mapping init;
+        for (auto [_, op] : links.front()) {
+            init[op.front()] = { op };
+        }
+        return build_map(init, links, transition);
+    }
+
+    op_mapping build_map(handle_t from, handle_t to, location_info_t &li) {
+        return build_map({gather_map(li, from.mod), gather_map(li, to.mod)}, li);
+    }
+
+    /* conversion_step::link_interface API */
+
+    operations conversion_step::children(operation) { VAST_UNIMPLEMENTED; }
+    operations conversion_step::children(operations) { VAST_UNIMPLEMENTED; }
+
+    operations conversion_step::parents(operation) { VAST_UNIMPLEMENTED;  }
+    operations conversion_step::parents(operations) { VAST_UNIMPLEMENTED; }
+
+    op_mapping conversion_step::parents_to_children() {
+        return reverse_mapping(children_to_parents());
+    }
+
+    op_mapping conversion_step::children_to_parents() {
+        return build_map(from(), to(), _location_info);
+    }
+
+    handle_t conversion_step::from() const { return _from; }
+    handle_t conversion_step::to() const { return _to; }
+
+    /* fat_link */
+
+    fat_link::fat_link(link_vector links)
+        : links(std::move(links)),
+          down(build_map(this->links)),
```

Comment:
this should be `_children` and `_parents` no?

Reply:
Forgot to update it here.